File: services/user-service/app/schemas/color.py
"""
AuraFit — Color Intelligence Pydantic schemas.
Request/response contracts for all color analysis endpoints.
"""
from __future__ import annotations

from enum import StrEnum
from uuid import UUID

from pydantic import Field

from app.schemas.base import AuraFitSchema, TimestampSchema, UUIDSchema


# ── Shared sub-types ──────────────────────────────────────────────────────────

class ColorEntry(AuraFitSchema):
    """A single color in a palette."""
    hex: str       = Field(..., description="6-digit hex e.g. '#C9A84C'")
    name: str      = Field(..., description="Human-readable color name")
    category: str  = Field(..., description="neutral|accent|base|statement")
    season_fit: str | None = Field(None, description="best|good|avoid")


class LipstickShade(AuraFitSchema):
    shade: str
    hex: str
    finish: str   = Field(..., description="matte|satin|gloss|sheer")
    intensity: str = Field(..., description="light|medium|bold")
    occasion: str  = Field(..., description="everyday|evening|professional|special")
    brands: list[str] = Field(default_factory=list)
    rationale: str


class HairColorOption(AuraFitSchema):
    name: str
    hex: str
    technique: str   = Field(..., description="all-over|highlights|balayage|ombre|gloss")
    commitment: str  = Field(..., description="permanent|semi-permanent|temporary")
    rationale: str
    maintenance: str = Field(..., description="Low|Medium|High")


class OutfitPalette(AuraFitSchema):
    occasion: str
    base_colors: list[ColorEntry]
    accent_colors: list[ColorEntry]
    avoid_colors: list[ColorEntry]
    styling_tip: str


class JewelryRecommendation(AuraFitSchema):
    metal_tones: list[str]      = Field(..., description="gold|rose-gold|silver|bronze|copper")
    gem_colors: list[ColorEntry]
    rationale: str
    style_notes: str


class MakeupColorRecommendation(AuraFitSchema):
    foundation_undertone: str
    foundation_finish: str
    blush_shades: list[ColorEntry]
    eyeshadow_palettes: list[dict]
    eyeliner_colors: list[ColorEntry]
    highlight_shades: list[ColorEntry]
    contour_shades: list[ColorEntry]


# ── Request schemas ───────────────────────────────────────────────────────────

class ColorAnalysisRequest(AuraFitSchema):
    """
    Compute color profile from a scan. If scan_id is omitted,
    uses the user's most recent active facial scan.
    """
    scan_id: UUID | None = Field(
        None,
        description="Facial scan to derive color profile from. Defaults to latest active scan."
    )
    use_extended_seasons: bool = Field(
        True,
        description="Whether to classify into the 12-season system (vs 4-season)"
    )


# ── Response schemas ──────────────────────────────────────────────────────────

class SeasonClassification(AuraFitSchema):
    season: str
    confidence: float = Field(ge=0.0, le=1.0)
    description: str
    key_characteristics: list[str]
    celebrity_examples: list[str]
    season_family: str   = Field(..., description="spring|summer|autumn|winter")
    contrast_level: str  = Field(..., description="low|medium|high")
    chroma_level: str    = Field(..., description="muted|soft|clear|bright")
    value_level: str     = Field(..., description="light|medium|deep")


class PersonalPalette(AuraFitSchema):
    best: list[ColorEntry]
    neutrals: list[ColorEntry]
    accents: list[ColorEntry]
    avoid: list[ColorEntry]
    hero_color: ColorEntry   = Field(..., description="Single best color for this person")


class ColorProfileOut(UUIDSchema, TimestampSchema):
    user_id: UUID
    scan_id: UUID | None
    season: str
    season_confidence: float
    season_description: str | None
    skin_tone: str | None
    undertone: str | None
    skin_hex: str | None
    ita_angle: float | None
    fitzpatrick: int | None
    hair_color_hex: str | None
    palette_best: list | None
    palette_avoid: list | None
    palette_neutrals: list | None
    palette_accents: list | None
    makeup_recommendations: dict | None
    lipstick_recommendations: list | None
    hair_color_recommendations: list | None
    outfit_recommendations: dict | None
    jewelry_recommendations: dict | None
    engine_version: str
    is_active: bool


class FullColorAnalysisResponse(AuraFitSchema):
    """Full response returned immediately on compute (or from cache)."""
    profile_id: UUID
    scan_id: UUID | None
    season: SeasonClassification
    palette: PersonalPalette
    makeup: MakeupColorRecommendation
    lipstick: list[LipstickShade]
    hair_color: list[HairColorOption]
    outfits: list[OutfitPalette]
    jewelry: JewelryRecommendation
    computed_at: str
    engine_version: str

File: apps/web/src/types/color.ts
// ── Color Intelligence Types (Stage 5) ──────────────────────────────────────
// Mirrors the backend Pydantic schemas in app/schemas/color.py

export type ColorSeasonFamily = "spring" | "summer" | "autumn" | "winter";

export type ColorSeason =
  | "spring" | "summer" | "autumn" | "winter"
  | "light_spring" | "true_spring" | "bright_spring"
  | "light_summer" | "true_summer" | "soft_summer"
  | "soft_autumn" | "true_autumn" | "deep_autumn"
  | "deep_winter" | "true_winter" | "bright_winter";

export type ContrastLevel = "low" | "medium" | "high";
export type ChromaLevel   = "muted" | "soft" | "clear" | "bright";
export type ValueLevel    = "light" | "medium" | "deep";

export interface ColorEntry {
  hex:        string;
  name:       string;
  category:   "neutral" | "accent" | "base" | "statement";
  season_fit?: "best" | "good" | "avoid";
}

export interface SeasonClassification {
  season:              ColorSeason;
  confidence:          number;      // 0–1
  description:         string;
  key_characteristics: string[];
  celebrity_examples:  string[];
  season_family:       ColorSeasonFamily;
  contrast_level:      ContrastLevel;
  chroma_level:        ChromaLevel;
  value_level:         ValueLevel;
}

export interface PersonalPalette {
  best:       ColorEntry[];
  neutrals:   ColorEntry[];
  accents:    ColorEntry[];
  avoid:      ColorEntry[];
  hero_color: ColorEntry;
}

export interface LipstickShade {
  shade:     string;
  hex:       string;
  finish:    "matte" | "satin" | "gloss" | "sheer" | "metallic";
  intensity: "light" | "medium" | "bold";
  occasion:  "everyday" | "evening" | "professional" | "special";
  brands:    string[];
  rationale: string;
}

export interface HairColorOption {
  name:        string;
  hex:         string;
  technique:   "all-over" | "highlights" | "balayage" | "ombre" | "gloss";
  commitment:  "permanent" | "semi-permanent" | "temporary";
  rationale:   string;
  maintenance: "Low" | "Medium" | "High";
}

export interface OutfitPalette {
  base_colors:   ColorEntry[];
  accent_colors: ColorEntry[];
  avoid_colors:  ColorEntry[];
  styling_tip:   string;
}

export interface JewelryRecommendation {
  metal_tones: string[];
  gem_colors:  ColorEntry[];
  rationale:   string;
  style_notes: string;
}

export interface MakeupColorRecommendation {
  foundation_undertone: string;
  foundation_finish:    string;
  blush_shades:         ColorEntry[];
  eyeshadow_palettes:   Array<{ name: string; shades: string[]; finish: string }>;
  eyeliner_colors:      ColorEntry[];
  highlight_shades:     ColorEntry[];
  contour_shades:       ColorEntry[];
}

export interface FullColorAnalysisResponse {
  profile_id:    string;
  scan_id:       string | null;
  season:        SeasonClassification;
  palette:       PersonalPalette;
  makeup:        MakeupColorRecommendation;
  lipstick:      LipstickShade[];
  hair_color:    HairColorOption[];
  outfits:       Record<string, OutfitPalette>;  // work|evening|casual|sport
  jewelry:       JewelryRecommendation;
  computed_at:   string;
  engine_version:string;
}

export interface ColorProfileRecord {
  id:              string;
  user_id:         string;
  scan_id:         string | null;
  season:          ColorSeason;
  season_confidence: number;
  season_description: string | null;
  skin_tone:       string | null;
  undertone:       string | null;
  skin_hex:        string | null;
  ita_angle:       number | null;
  fitzpatrick:     number | null;
  hair_color_hex:  string | null;
  palette_best:    Array<{ hex: string; name: string; category: string }> | null;
  palette_avoid:   Array<{ hex: string; name: string; category: string }> | null;
  palette_neutrals:Array<{ hex: string; name: string; category: string }> | null;
  palette_accents: Array<{ hex: string; name: string; category: string }> | null;
  makeup_recommendations:     MakeupColorRecommendation | null;
  lipstick_recommendations:   LipstickShade[] | null;
  hair_color_recommendations: HairColorOption[] | null;
  outfit_recommendations:     Record<string, OutfitPalette> | null;
  jewelry_recommendations:    JewelryRecommendation | null;
  engine_version:  string;
  is_active:       boolean;
  created_at:      string;
  updated_at:      string;
}

// Season UI metadata (used for theming / display)
export const SEASON_DISPLAY: Record<string, {
  label: string;
  family: ColorSeasonFamily;
  gradient: string;
  textColor: string;
  borderColor: string;
  emoji: string;
}> = {
  // Spring family
  true_spring:   { label: "True Spring",   family: "spring", gradient: "from-amber-400 to-orange-300",  textColor: "text-amber-900",  borderColor: "border-amber-400/40", emoji: "🌸" },
  light_spring:  { label: "Light Spring",  family: "spring", gradient: "from-yellow-300 to-pink-200",   textColor: "text-yellow-900", borderColor: "border-yellow-300/40",emoji: "🌷" },
  bright_spring: { label: "Bright Spring", family: "spring", gradient: "from-orange-400 to-yellow-300", textColor: "text-orange-900", borderColor: "border-orange-400/40",emoji: "✨" },
  spring:        { label: "Spring",        family: "spring", gradient: "from-amber-400 to-orange-300",  textColor: "text-amber-900",  borderColor: "border-amber-400/40", emoji: "🌸" },
  // Summer family
  true_summer:   { label: "True Summer",   family: "summer", gradient: "from-purple-400 to-pink-300",   textColor: "text-purple-900", borderColor: "border-purple-400/40",emoji: "🌿" },
  light_summer:  { label: "Light Summer",  family: "summer", gradient: "from-blue-300 to-purple-200",   textColor: "text-blue-900",   borderColor: "border-blue-300/40",  emoji: "💐" },
  soft_summer:   { label: "Soft Summer",   family: "summer", gradient: "from-slate-400 to-purple-300",  textColor: "text-slate-900",  borderColor: "border-slate-400/40", emoji: "🌫️" },
  summer:        { label: "Summer",        family: "summer", gradient: "from-purple-400 to-pink-300",   textColor: "text-purple-900", borderColor: "border-purple-400/40",emoji: "🌿" },
  // Autumn family
  true_autumn:   { label: "True Autumn",   family: "autumn", gradient: "from-orange-600 to-amber-500",  textColor: "text-orange-100", borderColor: "border-orange-500/40",emoji: "🍂" },
  soft_autumn:   { label: "Soft Autumn",   family: "autumn", gradient: "from-stone-500 to-amber-400",   textColor: "text-stone-100",  borderColor: "border-stone-400/40", emoji: "🍁" },
  deep_autumn:   { label: "Deep Autumn",   family: "autumn", gradient: "from-red-800 to-orange-700",    textColor: "text-red-100",    borderColor: "border-red-700/40",   emoji: "🌰" },
  autumn:        { label: "Autumn",        family: "autumn", gradient: "from-orange-600 to-amber-500",  textColor: "text-orange-100", borderColor: "border-orange-500/40",emoji: "🍂" },
  // Winter family
  true_winter:   { label: "True Winter",   family: "winter", gradient: "from-indigo-700 to-blue-600",   textColor: "text-indigo-100", borderColor: "border-indigo-500/40",emoji: "❄️" },
  bright_winter: { label: "Bright Winter", family: "winter", gradient: "from-blue-700 to-fuchsia-600",  textColor: "text-blue-100",   borderColor: "border-blue-500/40",  emoji: "💎" },
  deep_winter:   { label: "Deep Winter",   family: "winter", gradient: "from-slate-800 to-indigo-700",  textColor: "text-slate-100",  borderColor: "border-slate-600/40", emoji: "🌑" },
  winter:        { label: "Winter",        family: "winter", gradient: "from-indigo-700 to-blue-600",   textColor: "text-indigo-100", borderColor: "border-indigo-500/40",emoji: "❄️" },
};

export const SEASON_FAMILY_COLORS: Record<ColorSeasonFamily, { bg: string; text: string; border: string }> = {
  spring: { bg: "bg-amber-500/10",   text: "text-amber-400",  border: "border-amber-500/20"  },
  summer: { bg: "bg-purple-500/10",  text: "text-purple-400", border: "border-purple-500/20" },
  autumn: { bg: "bg-orange-500/10",  text: "text-orange-400", border: "border-orange-500/20" },
  winter: { bg: "bg-blue-500/10",    text: "text-blue-400",   border: "border-blue-500/20"   },
};
